### PR TITLE
Implement new "PowerNukkitOverworld" generator for future update, Implement better ore generator

### PR DESCRIPTION
**Pull request by GoodLucky777**
- Implement PowerNukkitOverworld generator for future update.
  - Implement PopulatorOreV2, and add some 1.17 block populator, add BlockTuff
- Deprecate PopulatorOreEmerald

Screenshot_Minecraft
Screenshot_Minecraft

~~This will be useful for generator something like deepslate ores~~ ✅ Added

### CODE REVIEW

**Comment by MonoAdrian23** [comment]
Line 213 Block Stone 
You have typed GRANITR

**Comment by GoodLucky777** [comment]
> Line 213 Block Stone
> You have typed GRANITR
> Use PopulatorOreV2 for all ore.... file

I already fixed it at next commit "fix"

**Comment by MonoAdrian23** [comment]
Ah i see it 
Sorry

**Review by LoboMetalurgico — CHANGES_REQUESTED** [review]
Hi, Thanks for your PR.

Talking to other developers, we thought it would be good to put an option that allows the server owner to switch between legacy generation and this one. This option would come to the new generation on new worlds, but with the legacy generation on existing worlds.

Also, if you can put the TUFF in another PR, I would be happy to approve the PR with the TUFF as soon as it opens. But remember to register it in Block.

**Comment by GoodLucky777** [comment]
Okay I will make another PR for Tuff

**Comment by LoboMetalurgico** [comment]
Thanks!

**Comment by GoodLucky777** [comment]
I also add deepslate ores to populator
Screenshot_Minecraft

**Comment by joserobjr** [comment]
Hey @GoodLuck777, we are really liking your changes to the populator. But we would like to give the owners the ability to choose between Cloudburst Nukkit generator and PowerNukkit generator, where the first would generate chunks exactly like the upstream does, to avoid the appearance of chunk borders, and the last would generate the world with all new biomes, blocks, and all cool stuff, and it would be the default for new worlds in PN, but for existing CBN worlds, the old generator is used until the owner changes it to the new generator.

You are more experience with worldgen than everybody on the team now lol do you think it is possible to do?

**Comment by GoodLucky777** [comment]
I register "PowerNukkitOverworld" generator for new update and revert changes of "Normal" generator

File: src/main/java/cn/nukkit/Server.java [hunk 1]
```diff
@@ -535,7 +536,7 @@ public Level remove(Object key) {
                 put("generator-settings", "");
                 put("level-name", "world");
                 put("level-seed", "");
-                put("level-type", "DEFAULT");
+                put("level-type", "PowerNukkitOverworld"); // Now we use PowerNukkit generator
```
Comment (alikindsys) [inline src/main/java/cn/nukkit/Server.java]: Using the new experimental generator should be opt-in.
```suggestion
                put("level-type", "DEFAULT");
```

**Review by alikindsys — CHANGES_REQUESTED** [review]

**Comment by alikindsys** [comment]
On `OreV2` and `PopulatorOreV2`
---
This PR adds **nothing new**.
It merely reverses the api from `PopulatorOre` from taking a block and the `OreType`s that are replaces it and turns it *upside down* by making the `OreV2`s themselves hold the information about what blocks are replaced by them under `ReplaceRules::mayReplace`.  
The code that generates the ores itself, [`PopulatorOreV2::generate`](https://github.com/PowerNukkit/PowerNukkit/blob/92695bcf85f10b5d45ce0864242a380a22a86938/src/main/java/cn/nukkit/level/generator/populator/impl/PopulatorOreV2.java#L44-L94) is copied verbatim from the old [`OreType::spawn`](https://github.com/PowerNukkit/PowerNukkit/blob/30019b85b251e84d3ba5ec3a6a0b3e01cc7712c1/src/main/java/cn/nukkit/level/generator/object/ore/OreType.java#L53-L101) with the subtle difference that it now doesn't return true since it's a void function.  
It proposes itself to be a "better generator" but does more of the same, and it doesn't do a good job on either of them. The API is way more obtuse due to the inversion done to the data propagation.  
Now every ore has to include the block it replaces and that duplicates data.  
Just think for yourselves if an ore needs to know what blocks it replaces or if the block being replaced should store which ores that are replaced by it.

Old API:
https://github.com/PowerNukkit/PowerNukkit/blob/30019b85b251e84d3ba5ec3a6a0b3e01cc7712c1/src/main/java/cn/nukkit/level/generator/Flat.java#L74-L83

New API:
https://github.com/PowerNukkit/PowerNukkit/blob/92695bcf85f10b5d45ce0864242a380a22a86938/src/main/java/cn/nukkit/level/generator/PowerNukkitOverworld.java#L75-L93

To say that this PR just shuffles things around would be an overstatement.

tl;dr: If you need to change the API to support that multiple blocks can be replaced by the same generators, change just that. Don't rewrite the whole wheel.

***Proposed Solution***: 
```java
class PopulatorOreMany extends Populator {
     private List<BlockState> replacedStates;
     private OreType[] oreTypes;

     @Override
     public void populate(ChunkManager level, int chunkX, int chunkZ, NukkitRandom random, FullChunk chunk) {
        int sx = chunkX << 4;
        int ex = sx + 15;
        int sz = chunkZ << 4;
        int ez = sz + 15;
        for (OreType type : this.oreTypes) {
            for (int i = 0; i < type.clusterCount; i++) {
                int x = NukkitMath.randomRange(random, sx, ex);
                int z = NukkitMath.randomRange(random, sz, ez);
                int y = NukkitMath.randomRange(random, type.minHeight, type.maxHeight);
                if (!replacedStates.contains(level.getBlockStateAt(x, y, z))) {
                    continue;
                }
            type.spawn(level, random, replaceId, x, y, z);
        }
    }
    /* Getters, setters and constructors omitted since they can be auto-generated */
}

```

List of mappings
---
[`OreType::blockInt`,`OreType::blockData`] ~> `OreV2.ReplaceRule::placesBlock`  
`OreType::clusterCount` -> `PopulatorOreV2::iterations` 
`OreType::clusterSize` -> `OreV2::count`  
`OreType::maxHeight` -> `PopulatorOreV2::endY`  
`OreType::minHeight` -> `PopulatorOreV2::startY`  
`OreType::replaceBlockId` ~> `OreV2.ReplaceRules::mayReplace`  

`PopulatorOre::replaceId` -> `OreV2.ReplaceRule::placesBlock`
`PopulatorOre::oreTypes` ~> `OreV2.ReplaceRule::mayReplace`

The only things that are left unmapped are `PopulatorOreV2::startX`, `PopulatorOreV2::endX`, `PopulatorOreV2::startZ` and `PopulatorOreV2::endZ`, all of which makes no sense since ore generation in Minecraft isn't bound on the X or Z axis, there only exists limitations on the spawn height (Y axis).

File: src/main/java/cn/nukkit/Server.java (hunk 1 above)
Comment (GoodLucky777) [reply src/main/java/cn/nukkit/Server.java]: Okay

**Comment by GoodLucky777** [comment]
The original intention was to make it similar to the Behavior Pack's Features, Feature rules syntax of Bedrock Edition, so I tried to make it like that, but I put it on hold because the official 1.18 release hasn't been made yet. If I keep only this function in this PR without adding more feature to that class, I'll add it to the PopulatorOre class. I almost forgot this problem. Thank you for letting me know this

**Comment by GoodLucky777** [comment]
Closing this PR because PN is no longer maintained.
